lockup/src/lib.rs

```diff
@@ -1662,6 +1662,317 @@ mod tests {
+
+        // All the tokens are still locked
+        assert_eq!(contract.get_owners_balance().0, 0);
+        assert_eq!(contract.get_liquid_owners_balance().0, 0);
+        assert_eq!(contract.get_locked_amount().0, lockup_amount);
+
+        // Nothing new is vested since termination
+        let unvested_amount_day_before_lockup_cliff =
+            contract.get_unvested_amount(vesting_schedule.clone()).0;
+        assert_eq!(
+            unvested_amount_day_before_lockup_cliff,
+            unvested_amount_at_termination_day
+        );
+
+        // *** day 865: day of lockup cliff ***
+        let ts_day_of_lockup_cliff = to_ts(GENESIS_TIME_IN_DAYS + YEAR);
+        context.block_timestamp = ts_day_of_lockup_cliff;
+        testing_env!(context.clone());
+
+        let unlocked_amount_day_of_lockup_cliff = contract.get_liquid_owners_balance().0;
+        let expected_unlocked_amount_day_of_lockup_cliff = lockup_amount
+            / (contract.lockup_information.release_duration.unwrap() as u128)
+            * (to_nanos(YEAR) as u128);
+        assert_almost_eq_with_max_delta(
+            expected_unlocked_amount_day_of_lockup_cliff,
+            unlocked_amount_day_of_lockup_cliff,
+            to_yocto(1),
+        );
+        assert_eq!(to_yocto(250000), unlocked_amount_day_of_lockup_cliff);
+
+        // *** day 1064: 1 day before unlock stopped ***
+        let ts_1_day_before_unlock_stop = to_ts(GENESIS_TIME_IN_DAYS + YEAR + 199);
+        context.block_timestamp = ts_1_day_before_unlock_stop;
+        testing_env!(context.clone());
+
+        let locked_amount_1_day_before_unlock_stop = contract.get_locked_amount().0;
+
+        // *** day 1065, unlock stopped ***
+        let ts_day_unlock_stopped = to_ts(GENESIS_TIME_IN_DAYS + YEAR + 200);
+        context.block_timestamp = ts_day_unlock_stopped;
+        testing_env!(context.clone());
+
+        let locked_amount_day_unlock_stopped = contract.get_locked_amount().0;
+        assert!(
+            locked_amount_day_unlock_stopped < locked_amount_1_day_before_unlock_stop,
+            "Locked amount should decrease"
+        );
+        assert_eq!(
+            locked_amount_day_unlock_stopped,
+            unvested_amount_at_termination_day
+        );
+
+        // We unlock 684 tokens in one day,
+        // it means that our delta (1 token) is small enough to detect any issue
+        assert_almost_eq_with_max_delta(
+            to_yocto(684),
+            locked_amount_1_day_before_unlock_stop - locked_amount_day_unlock_stopped,
+            to_yocto(1),
+        );
+
+        // *** day 1066, 1 day after unlock stopped ***
+        let ts_1_day_after_unlock_stopped = to_ts(GENESIS_TIME_IN_DAYS + YEAR + 201);
+        context.block_timestamp = ts_1_day_after_unlock_stopped;
+        testing_env!(context.clone());
+
+        assert_eq!(
+            contract.get_locked_amount().0,
+            unvested_amount_at_termination_day
+        );
```

Comment:
Please, add assert on `get_unlocked() == initial lockup amount - terminated`

Reply:
Sorry, I will not add this check because both possible methods are broken
https://github.com/near/core-contracts/issues/192#issuecomment-996766606
I guess they will work for this specific case, but anyway, I don't want to use them, that's a matter of coincidence that they give us the expected result here

P.S. It's still too early to review this; I will ping you separately when I finish the work here

Reply:
As per our last discussion, let's add this assert anyway, and if it crashes (supposedly in the next test), let's comment it out and leave an explanation why we believe it is not critical